src/App.js

```diff
@@ -14,35 +13,53 @@ class App extends Component {
     super(props);
     this.state = {
       errorMessage: null,
+      transactionState: {
+        transactions: [],
+        sorted: {},
+      },
     };
   }
 
   setErrorMessage = (message) => this.setState({ errorMessage: message });
 
+  addTransactionToTransactionState = (transaction) =>
+    this.setState((state) => {
+      let newTransactions = [...state.transactionState.transactions];
+      newTransactions.push(transaction);
+      return {
+        transactionState: {
+          transactions: newTransactions,
+          sorted: { ...state.transactionState.sorted },
+        },
+      };
+    });
+
   render() {
     return (
-      <TransactionState>
-        <CalculationState>
-          <Fragment>
-            <h1 className="center-align">Crypto Capital Gains Calculator</h1>
-            <div className="container">
-              <p className="flow-text center">
-                This app was built mainly to get practice designing and building
-                React apps. I am not an accountant and make no guarantees that
-                the calculations performed within this app are accurate. I
-                encourage you to seek the help of an accountant or tax expert
-                with any questions regarding tax obligations for crypto
-                transactions.
-              </p>
-            </div>
-            <div className="divider"></div>
-            <TransactionForm />
-            <TransactionTable setErrorMessage={this.setErrorMessage} />
-            <ErrorMessage errorMessage={this.state.errorMessage} />
-            <CalculationTable />
-          </Fragment>
-        </CalculationState>
-      </TransactionState>
+      <CalculationState>
+        <Fragment>
+          <h1 className="center-align">Crypto Capital Gains Calculator</h1>
+          <div className="container">
+            <p className="flow-text center">
+              This app was built mainly to get practice designing and building
+              React apps. I am not an accountant and make no guarantees that the
+              calculations performed within this app are accurate. I encourage
+              you to seek the help of an accountant or tax expert with any
+              questions regarding tax obligations for crypto transactions.
+            </p>
+          </div>
+          <div className="divider"></div>
+          <TransactionForm
+            addTransaction={this.addTransactionToTransactionState}
+          />
+          <TransactionTable
+            setErrorMessage={this.setErrorMessage}
+            transactions={this.state.transactionState.transactions}
```

Comment:
this may be breaking encapsulation by exposing the internal `transactions` array. maybe use `transactions.slice()`?

Reply:
The only reason I could think for passing in the actual state is if the component receiving it has a need to re-render based on a state change, but I believe updating a prop to a component also triggers a re-rendering, so the .slice or similar method should still trigger an update.

For consistency, I'm going to stick with the spread operator instead of using the slice method, but the end result will be the same (copy of the array instead of a pointer to the real thing)

Reply:
This would also be true for the ErrorMessage component which receives this.state.errorMessage as a prop